db/transaction/retryable_transaction.go

```diff
@@ -0,0 +1,171 @@
+package transaction
+
+import (
+	"time"
+
+	"strings"
+
+	"github.com/cloudwan/gohan/db/pagination"
+	"github.com/cloudwan/gohan/schema"
+	"github.com/cloudwan/gohan/util"
+	"github.com/jmoiron/sqlx"
+)
+
+const NUMBER_OF_RETRIES_CONFIG_KEY = "database/transaction_retries/attempts"
+const RETRY_STRATEGY_CONFIG_KEY = "database/transaction_retries/strategy"
+const INTERVAL_BETWEEN_ATTEMPTS_CONFIG_KEY = "database/transaction_retries/interval_between_attempts"
+const DEFAULT_NUMBER_OF_RETRIES = 1
+const DEFAULT_RETRY_STRATEGY = "deadlock"
+const DEFAULT_INTERVAL_BETWEEN_ATTEMPTS = "100ms"
+
+const MYSQL_DEADLOCK_MSG = "Error 1213: Deadlock found when trying to get lock; try restarting transaction"
+const SQLITE_DEADLOCK_MSG = "database is locked"
+
+type RetryConfig struct {
+	Attempts      int
+	Strategy      RetryStrategyPredicate
+	SleepInterval time.Duration
+}
+
+var cache RetryConfig
+var cacheInitialized = false
+
+type RetryableTransaction struct {
+	Tx     Transaction
+	Config RetryConfig
+}
+
+type RetryStrategyPredicate func(error) bool
+
+var strategies = map[string]RetryStrategyPredicate{
+	"deadlock": IsDeadlock,
+}
+
+func NewRetryableTransaction(transaction Transaction) Transaction {
+	config := readConfig()
+	tx := &RetryableTransaction{
+		Tx:     transaction,
+		Config: config,
+	}
+	log.Debug("Created retryable transaction tx: %v, rawTx: %v, attempts: %d, interval: %s", &tx, transaction.RawTransaction(), config.Attempts, config.SleepInterval)
+	return tx
+}
+
+func IsDeadlock(err error) bool {
+	return strings.Contains(err.Error(), MYSQL_DEADLOCK_MSG) || strings.Contains(err.Error(), SQLITE_DEADLOCK_MSG)
+}
+
+func readConfig() RetryConfig {
+	if cacheInitialized {
+		return cache
+	}
+	config := util.GetConfig()
+	attempts := config.GetInt(NUMBER_OF_RETRIES_CONFIG_KEY, DEFAULT_NUMBER_OF_RETRIES)
+	configStrategy := config.GetString(RETRY_STRATEGY_CONFIG_KEY, DEFAULT_RETRY_STRATEGY)
+	retryStrategy, ok := strategies[configStrategy]
+	if !ok {
```

Comment:
this should err for good. Also, the default should be applied when strategy is missing (or is empty string, because it's easier).